osgp/protocol-adapter-iec60870/osgp-protocol-adapter-iec60870/src/main/java/org/opensmartgridplatform/adapter/protocol/iec60870/application/services/LightMeasurementGatewayDeviceResponseService.java

```diff
@@ -0,0 +1,94 @@
+
+@Service
+public class LightMeasurementGatewayDeviceResponseService extends AbstractDeviceResponseService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LightMeasurementGatewayDeviceResponseService.class);
+
+    private static final DeviceType DEVICE_TYPE = DeviceType.LIGHT_MEASUREMENT_GATEWAY;
+
+    @Autowired
+    private Iec60870DeviceRepository iec60870DeviceRepository;
+
+    @Autowired
+    private ResponseMetadataFactory responseMetadataFactory;
+
+    @Autowired
+    private LightMeasurementService lightMeasurementService;
+
+    public LightMeasurementGatewayDeviceResponseService() {
+        super(DEVICE_TYPE);
+    }
+
+    @Override
+    public void process(final MeasurementReportDto measurementReportDto, final ResponseMetadata responseMetadata) {
+        LOGGER.info("Received measurement report {} for light measurement gateway {}.", measurementReportDto,
+                responseMetadata.getDeviceIdentification());
+
+        final ResponseMetadata newResponseMetadata = this.responseMetadataFactory
+                .createWithNewCorrelationUid(responseMetadata);
+
+        final List<Iec60870Device> lightMeasurementDevices = this.iec60870DeviceRepository
+                .findByGatewayDeviceIdentification(responseMetadata.getDeviceIdentification());
+
+        this.sendLightSensorStatusResponses(measurementReportDto, lightMeasurementDevices, newResponseMetadata);
+    }
+
+    private void sendLightSensorStatusResponses(final MeasurementReportDto measurementReportDto,
+            final List<Iec60870Device> lightMeasurementDevices, final ResponseMetadata responseMetadata) {
+
+        for (final MeasurementGroupDto mg : measurementReportDto.getMeasurementGroups()) {
+            final Optional<Iec60870Device> device = lightMeasurementDevices.stream()
+                    .filter(d -> d.getInformationObjectAddress().toString().equals(mg.getIdentification()))
+                    .findFirst();
+            device.ifPresent(d -> this.sendLightSensorStatusResponse(mg, d, responseMetadata));
+        }
+    }
+
+    private void sendLightSensorStatusResponse(final MeasurementGroupDto measurementGroupDto,
+            final Iec60870Device device, final ResponseMetadata responseMetadata) {
+        final ResponseMetadata rm = new ResponseMetadata.Builder()
+                .withCorrelationUid(responseMetadata.getCorrelationUid())
+                .withDeviceIdentification(device.getDeviceIdentification())
+                .withDomainInfo(device.getDeviceType().domainType().domainInfo())
+                .withMessageType(MessageType.GET_LIGHT_SENSOR_STATUS.name())
+                .withOrganisationIdentification(responseMetadata.getOrganisationIdentification())
+                .build();
+
+        final boolean on = ((BitmaskMeasurementElementDto) measurementGroupDto.getMeasurements()
+                .get(0)
+                .getMeasurementElements()
+                .get(0)).getValue() == 1;
```

Comment:
In the case of light measurement gateway/devices, the result of a general interrogation of the gateway device should result in information objects for each light measurement  device, which in turn is represented as a single-point information element. So for a general interrogation each light measurement device will be represented by a measurement group with a single measurement containing a single measurement element (in the form of a BitmaskMeasurementElementDto).
For events this will be different, as they will also contain a timestamp.
I would expect events to be handled by the LightMeasurementDeviceResponseService though.
Refactored the code.